Index: HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.view.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.view.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.view.js	(revision )
@@ -4,6 +4,7 @@
     APP.view.NewsList = APP.View.$extend({
         container: "#newsList",
         selectItem: function (itemId) {
+            console.log("view selectItem", itemId)
             if (this._$selectedItem) {
                 this._$selectedItem.removeClass("selected");
             }
\ No newline at end of file
Index: HTML5Zen/test/mockServices/NewsList.service.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/test/mockServices/NewsList.service.js	(revision )
+++ HTML5Zen/test/mockServices/NewsList.service.js	(revision )
@@ -0,0 +1,28 @@
+(function () {
+    "use strict";
+
+    APP.service.NewsList = APP.service.NewsList.$extend({
+        fetch: function () {
+            console.log("called");
+            var that = this;
+            that.updated(getMockData());
+        }
+    });
+
+    function getMockData() {
+        return (function () {
+            var items = [];
+            var n = 0;
+            while (n < 20) {
+                items.push({
+                    id: n,
+                    title: "News " + n,
+                    description: "Description " + n + ": " + new Date().toLocaleTimeString() + " lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum"
+                });
+                n++;
+            }
+            return items;
+        }());
+    }
+}());
+
Index: HTML5Zen/code/lib/js/app/modules/News/News.controller.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/News.controller.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/modules/News/News.controller.js	(revision )
@@ -9,58 +9,35 @@
      * @param {APP.controller.NewsDetail} newsDetail
      */
     APP.controller.News = APP.Controller.$extend({
-        init: function (options, service, newsList, newsDetail) {
-            this.options = options;
-            this.service = service;
-            this.newsList = newsList;
-            this.newsDetail = newsDetail;
+        getControls: function () {
+            return {
+                newsList: new APP.controller.NewsList(),
+                newsDetail: new APP.controller.NewsDetail()
+            };
         },
-        load: function () {
+        getServices: function () {
+            return {
+                newsList: new APP.service.NewsList(),
+                newsDetail: new APP.service.NewsDetail()
+            };
+        },
+        onStateChange: function (stateData) {
             var that = this;
-            var newsData;
-            var newsList, newsDetail, service;
-
-            APP.messaging.subscribe("appStateChange-view-changedTo-example", function () {
-                if (!that.loaded) {
-                    newsList = that.newsList || new APP.controller.NewsList({
-                        onNewsItemSelected: loadNewsDetails
-                    });
-                    newsDetail = that.newsDetail || new APP.controller.NewsDetail();
-                    service = that.service || new APP.service.News();
-                    service.fetch(function (data) {
-                        newsData = data;
-                        newsList.load(data);
-                    });
-                    that.loaded = true;
+            if (stateData.moduleItem) {
+                console.log("called onStateChange", stateData)
+                that.controls.newsList.selectItem(stateData.moduleItem);
+                that.services.newsDetail.fetch(that.services.newsList.lastResult(), stateData.moduleItem);
-                }
+            }
-            });
-
-            APP.messaging.subscribe("appStateChange-view-stateChange-example", function (pageData) {
-                if (pageData.moduleItem) {
-                    newsList.selectItem(pageData.moduleItem);
-                    loadNewsDetails(pageData.moduleItem);
-                }
-            });
-
-            /*this destroy is for keeping performance when there are multiple complex modules and you want to make the dom tree less heavy*/
-            APP.messaging.subscribe("appStateChange-view-changedFrom-example", function () {
-                newsList.destroy();
-                newsDetail.destroy();
-                that.destroy();
-            });
-
-            function loadNewsDetails(selectedItemId) {
-                newsDetail.load(APP.DOM.grep(newsData, function (item) {
-                    return String(item.id) === String(selectedItemId);
-                })[0]);
-            }
         },
-        destroy: function () {
-            this.$super();
-            this.service = null;
-            this.loaded = false;
-            this.newsList = null;
-            this.newsDetail = null;
+        messages: {
+            show: "appStateChange-view-changedTo-example",
+            hide: "appStateChange-view-changedFrom-example",
+            stateChange: "appStateChange-view-stateChange-example"
+//            ,others: {
+//                "message-name": function() {
+//                    this.onItemReceive();
+//                }
+//            }
         }
     });
 
\ No newline at end of file
Index: HTML5Zen/test/mockServices/News.service.js
===================================================================
--- HTML5Zen/test/mockServices/News.service.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/test/mockServices/News.service.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
@@ -1,26 +0,0 @@
-(function () {
-    "use strict";
-
-    APP.service.News = APP.service.News.$extend({
-        fetch: function (callback) {
-            callback(getMockData());
-        }
-    });
-
-    function getMockData() {
-        return (function () {
-            var items = [];
-            var n = 0;
-            while (n < 20) {
-                items.push({
-                    id: n,
-                    title: "News " + n,
-                    description: "Description " + n + " lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum"
-                });
-                n++;
-            }
-            return items;
-        }());
-    }
-}());
-
Index: HTML5Zen/code/lib/js/app/modules/News/News.service.js
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/News.service.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/modules/News/News.service.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
@@ -1,10 +0,0 @@
-(function ($) {
-    "use strict";
-
-    APP.service.News = APP.Service.$extend({
-        fetch: function (onSuccess) {
-            $.get(APP.serviceURL.news.fetch, onSuccess);
-        }
-    });
-
-}(APP.DOM));
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.controller.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.controller.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.controller.js	(revision )
@@ -8,11 +8,15 @@
      * @param {APP.Template=} template
      */
     APP.controller.NewsDetail = APP.Controller.$extend({
-        init: function (options, view, template) {
-            this.options = options;
-            this.template = template || new APP.Template("newsDetailTemplate", APP.Template.LOAD.BY_ID);
-            this.view = view || new APP.view.NewsDetail();
-            this.load({});
+        getTemplates: function () {
+            return {
+                newsDetail: new APP.Template("newsDetailTemplate")
+            };
+        },
+        getViews: function () {
+            return {
+                newsDetail: new APP.view.NewsDetail()
+            };
         }
     });
 
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.service.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.service.js	(revision )
+++ HTML5Zen/code/lib/js/app/modules/News/controls/newsDetail/NewsDetail.service.js	(revision )
@@ -0,0 +1,17 @@
+(function ($) {
+    "use strict";
+
+    APP.service.NewsDetail = APP.Service.$extend({
+        fetch: function (newsList, selectedItemId) {
+            if (newsList) {
+                var details = $.grep(newsList, function (item) {
+                    return String(item.id) === String(selectedItemId);
+                })[0];
+                this.updated(details);
+            } else {
+                this.updated({});
+            }
+        }
+    });
+
+}(APP.DOM));
\ No newline at end of file
Index: HTML5Zen/code/index.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/index.html	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/index.html	(revision )
@@ -113,6 +113,7 @@
     <script src="lib/js/vendor/handlebars-v1.1.2.js"></script>
 
     <script src="lib/js/app/namespace.js"></script>
+    <script src="lib/js/app/serviceURLs.js"></script>
 
     <script src="lib/js/app/utilities/DOM.js"></script>
     <script src="lib/js/app/utilities/classExtend.js"></script>
@@ -132,9 +133,10 @@
     <script src="lib/js/app/modules/News/controls/newsList/NewsList.controller.js"></script>
     <script src="lib/js/app/modules/News/controls/newsList/NewsList.view.js"></script>
     <script src="lib/js/app/modules/News/News.controller.js"></script>
-    <script src="lib/js/app/modules/News/News.service.js"></script>
+    <script src="lib/js/app/modules/News/controls/newsList/NewsList.service.js"></script>
+    <script src="lib/js/app/modules/News/controls/newsDetail/NewsDetail.service.js"></script>
 
-    <script src="../test/mockServices/News.service.js"></script>
+    <script src="../test/mockServices/NewsList.service.js"></script>
 
     <script src="lib/js/app/main.js"></script>
     <!--/MERGE SCRIPTS-->
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.service.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.service.js	(revision )
+++ HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.service.js	(revision )
@@ -0,0 +1,8 @@
+(function ($) {
+    "use strict";
+
+    APP.service.NewsList = APP.Service.$extend({
+        dataURL: APP.serviceURL.news.fetch
+    });
+
+}(APP.DOM));
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/utilities/baseClass/Controller.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/utilities/baseClass/Controller.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/utilities/baseClass/Controller.js	(revision )
@@ -1,27 +1,155 @@
 (function () {
     "use strict";
 
+    /**
+     *
+     * @param {Object=} options
+     * @param {APP.View=} views
+     * @param {APP.Template=} templates
+     * @param {APP.Service=} services
+     * @param {Function=} controls
+     */
     var Controller = Classy.$extend({
-        init: function (options, view, template) {
-            this.view = view;
-            this.template = template;
+        boolOnHideDestroy: false,
+        init: function (options, views, templates, services, controls) {
+            console.log("init controller")
+            this.views = views;
+            this.templates = templates;
             this.options = options;
+            this.services = services;
+            this.controls = controls;
         },
-        render: function (data) {
-            this.view.render(this.template.process(data));
+        getViews: function () {
+            return this.views;
         },
+        getTemplates: function () {
+            return this.templates;
+        },
+        getControls: function () {
+            return this.controls;
+        },
+        getServices: function () {
+            return this.service;
+        },
+        render: function (data, viewName) {
+            console.log("render", viewName, data)
+            this.views[viewName].render(this.templates[viewName].process(data));
+        },
         load: function (data) {
+            console.log("load", this._loaded);
             var that = this;
-            that.template.load(function () {
-                that.render(data);
-            });
+            that.views = that.views || that.getViews();
+            that.templates = that.templates || that.getTemplates();
+            that.controls = that.controls || that.getControls();
+            that.services = that.services || that.getServices();
+
+            var templates = this.templates, template;
+            if (!that._loaded) {
+                if (templates) {
+                    var renderFunction = function (template) {
+                        return function () {
+                            that.render(data, template);
+                        };
+                    };
+                    for (template in templates) {
+                        if (templates.hasOwnProperty(template)) {
+                            templates[template].load(renderFunction(template));
+                        }
+                    }
+                }
+
+                if (that.messages) {
+                    var messages = that.messages;
+                    APP.messaging.subscribe(messages.show, proxy(that, this.onShow));
+                    APP.messaging.subscribe(messages.hide, proxy(that, this.onHide));
+                    APP.messaging.subscribe(messages.stateChange, proxy(that, this.onStateChange));
+                }
+                that._loaded = true;
+            } else {
+                console.log("load _loaded")
+                for (template in templates) {
+                    if (templates.hasOwnProperty(template)) {
+                        that.render(data, template);
+                    }
+                }
+            }
         },
+        onStateChange: function (stateData) {
+        },
+        onUpdateReceive: function (data, item) {
+            console.log("onUpdateReceive", item, data)
+            var that = this;
+            that.controls[item].load(data);
+        },
         destroy: function () {
-            if (this.view) {
-                this.view.destroy();
+            if (this.messages) {
+                var messages = this.messages;
+                APP.messaging.unsubscribe(messages.show);
+                APP.messaging.unsubscribe(messages.hide);
+                APP.messaging.unsubscribe(messages.stateChange);
             }
+            if (this.views) {
+                loop(this.views, "destroy");
-        }
+            }
+            if (this.controls) {
+                loop(this.controls, "destroy");
+            }
+            if (this.services) {
+                loop(this.services, "destroy");
+            }
+            this.services = null;
+            this.views = null;
+            this.templates = null;
+            this.options = null;
+            this.controls = null;
+            this._loaded = false;
+            this._exist = false;
+        },
+        onShow: function () {
+            console.log("onShow controller")
+            if (!this._exist) {
+                initServices(this.services, this);
+                this._exist = true;
+            }
+        },
+        onHide: function () {
+            loop(this.services, "stop");
+            if (this.boolOnHideDestroy) {
+                this.destroy();
+            }
+        }
     });
+
+    function loop(collection, method, data) {
+        var item;
+        for (item in collection) {
+            if (collection.hasOwnProperty(item)) {
+                collection[item][method](data);
+            }
+        }
+    }
+
+    function proxy(scope, method) {
+        return function () {
+            method.apply(scope, arguments);
+        };
+    }
+
+    function initServices(services, that) {
+        var service;
+        for (service in services) {
+            if (services.hasOwnProperty(service)) {
+                services[service].onUpdate(getFunction(that, service));
+                services[service].fetch();
+            }
+        }
+    }
+
+    function getFunction(scope, service) {
+        return function (data) {
+            scope.onUpdateReceive(data, service);
+        };
+    }
 
     APP.Controller = Controller;
 
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/utilities/baseClass/Service.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/utilities/baseClass/Service.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/utilities/baseClass/Service.js	(revision )
@@ -1,15 +1,49 @@
-(function () {
+(function ($) {
     "use strict";
 
     var noop = function () {
     };
 
+    var privateVar = {};
+
     var Service = Classy.$extend({
-        fetch: noop,
+        dataURL: null,
+        fetch: function () {
+            $.get(this.dataURL, function (data) {
+                this.updated(data);
+            });
+        },
         update: noop,
         get: noop,
-        on: noop
+        lastResult: function () {
+            return privateVar[this._id_].lastResult;
+        },
+        onUpdate: function (callback) {
+            console.log("onUpdate service")
+            privateVar[this._id_].updateCallBack.push(callback);
+        },
+        init: function () {
+            console.log("init service")
+            this._id_ = (new Date()).getTime() + Math.floor(Math.random() * 1000000);
+            privateVar[this._id_] = {};
+            privateVar[this._id_].updateCallBack = [];
+        },
+        updated: function (data) {
+            var n, l;
+            var privateVar2 = privateVar[this._id_];
+            if (privateVar2) {
+                for (n = 0, l = privateVar2.updateCallBack.length; n < l; n++) {
+                    privateVar2.updateCallBack[n](data);
+                }
+            }
+            privateVar2.lastResult = data;
+        },
+        destroy: function () {
+            console.log("destroy service")
+            delete privateVar[this._id_];
+        },
+        stop: noop
     });
 
     APP.Service = Service;
-}());
\ No newline at end of file
+}(APP.DOM));
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.controller.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.controller.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/modules/News/controls/newsList/NewsList.controller.js	(revision )
@@ -2,16 +2,19 @@
     "use strict";
 
     APP.controller.NewsList = APP.Controller.$extend({
-        init: function (options, view, template) {
-            this.options = options;
-            this.view = view || new APP.view.NewsList();
-            this.template = template || new APP.Template("newsListTemplate");
+        getTemplates: function () {
+            return {
+                newsList: new APP.Template("newsListTemplate")
+            };
         },
-        render: function (data) {
-            this.view.render(this.template.process(data));
+        getViews: function () {
+            return {
+                newsList: new APP.view.NewsList()
+            };
         },
         selectItem: function (id) {
-            this.view.selectItem(id);
+            console.log("selectedItem", id)
+            this.views.newsList.selectItem(id);
         }
     });
 
\ No newline at end of file
Index: HTML5Zen/code/lib/js/app/main.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/main.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/main.js	(revision )
@@ -1,7 +1,7 @@
 (function () {
     "use strict";
 
-    new APP.controller.News().load();
+    (new APP.controller.News()).load();
     APP.navigation.load("introduction");
 
 }());
Index: HTML5Zen/code/lib/js/app/utilities/navigation.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- HTML5Zen/code/lib/js/app/utilities/navigation.js	(revision 76663139e78a18314f94186fce3a5ddda26a921b)
+++ HTML5Zen/code/lib/js/app/utilities/navigation.js	(revision )
@@ -2,6 +2,7 @@
     "use strict";
 
     var currentView;
+    var viewState = {};
 
     var Navigation = function () {
 
@@ -20,22 +21,29 @@
                 hidePage(currentView, appStateData);
                 currentView = appStateData.page;
                 showPage(currentView, appStateData);
-            } else {
-                APP.messaging.publish("appStateChange-view-stateChange-" + currentView, appStateData);
             }
+            viewState[currentView] = appStateData;
+            APP.messaging.publish("appStateChange-view-stateChange-" + currentView, appStateData);
             APP.messaging.publish("appStateChange", appStateData);
         }
 
         function hidePage(page, appStateData) {
             if (page) {
-                $("a[href^='#/" + page + "']").removeClass("selected");
+                $("a[href$='#/" + page + "']").removeClass("selected");
                 $("#" + page).removeClass("page-visible");
                 APP.messaging.publish("appStateChange-view-changedFrom-" + page, appStateData);
             }
         }
+
         function showPage(page, appStateData) {
             $("#" + page).addClass("page-visible");
-            $("a[href^='#/" + page + "']").addClass("selected");
+            $("a[href$='#/" + page + "']").addClass("selected");
+            var cachedViewState = viewState[page];
+            if (cachedViewState) {
+                if (cachedViewState.moduleItem) {
+                    location.replace("#/" + page + "/" + cachedViewState.module + "/" + cachedViewState.moduleItem);
+                }
+            }
             APP.messaging.publish("appStateChange-view-changedTo-" + currentView, appStateData);
         }
 
\ No newline at end of file
